src/main/java/de/tum/cit/aet/artemis/programming/service/ci/DependencyDownloadScript.java

```diff
@@ -0,0 +1,72 @@
+package de.tum.cit.aet.artemis.programming.service.ci;
+
+public enum DependencyDownloadScript {
```

Comment:
I don’t think this is going to work. `mvn install` will only download the dependencies required for build+test. Dependencies required for static code analysis or ones required by other more special plugins that are not executed during `mvn install` will still be missing and therefore break the build plan in case the network is disabled. You do not know which plugins (or Maven tasks) are going to be used in an exercise and can never statically pre-determine that.
I assume other build tools behave similarly regarding the lazy loading of additional plugins and dependencies.

More importantly from a user perspective:
I think that introducing a ‘magic’ invisible step that downloads dependencies before the actual build script that is visible to the instructor is quite counter-intuitive and will lead to more questions asked by instructors since they do no longer understand what is actually happening. I would like to see exactly which container image is started with which flags and which script is executed inside with as little ‘magic‘ surrounding it as possible. I.e. if the network is enabled, the script can just download them. If it is disabled, they must be part of the container.